File: 23-merge-k-sorted-lists/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

const MIN = 10001;

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
  const length = lists.length;
  let head: ListNode | null = null;
  let prev: ListNode | null = null;

  while (true) {
    let pos = -1;
    let val = MIN;
    for (let i = 0; i < length; i++) {
      if ((lists[i]?.val ?? MIN + 1) < val) {
        pos = i;
        val = lists[i]?.val as number;
      }
    }

    if (pos === -1) break;

    const now = new ListNode(val, null);
    if (!!prev) prev.next = now;
    else head = now;
    prev = now;

    lists[pos] = lists[pos].next;
  }

  return head;
}

File: 23-merge-k-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int n = lists.size();
        ListNode* head = NULL;
        ListNode* prev = NULL;
        while(true) {
            int minVal = INT_MAX, minIdx = -1;
            for(int i = 0; i < n; i++) {
                if(lists[i] != NULL && lists[i]->val < minVal) {
                    minVal = lists[i]->val;
                    minIdx = i;
                }
            }
            if(minIdx == -1) break;
            ListNode* node = new ListNode(minVal);
            if(head == NULL) head = node;
            if(prev != NULL) prev->next = node;
            lists[minIdx] = lists[minIdx]->next;
            prev = node;
        }
        return head;
    }
};
